### PR TITLE
spks automated billing

### PR DESCRIPTION
**Pull request by wejdross**
Add automated billing for spks. This code is intended to run as deployment and execute once per 24h. UoM is set to correspond:  https://docs.central.vshn.ch/unit-of-measurements.html#_uoms_for_metered_billing

### CODE REVIEW

File: pkg/cmd/spks.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+				EnvVars: []string{"SALES_ORDER"}, Destination: &salesOrder, Required: false, DefaultText: defaultTextForOptionalFlags, Value: "S10121"},
+			&cli.StringFlag{Name: "prometheus-url", Usage: "URL of the Prometheus API",
+				EnvVars: []string{"PROMETHEUS_URL"}, Destination: &prometheusURL, Required: true, DefaultText: defaultTextForRequiredFlags, Value: "http://prometheus-monitoring-application.monitoring-application.svc.cluster.local:9090"},
+			&cli.StringFlag{Name: "unit-id", Usage: "Unit ID for the consumed units",
+				EnvVars: []string{"UNIT_ID"}, Destination: &UnitID, Required: true, DefaultText: defaultTextForRequiredFlags, Value: "uom_uom_68_b1811ca1"},
+		},
+		Action: func(c *cli.Context) error {
+			// this function is intended to run once a day
+			// it will query the prometheus metrics for the last 24 hours
+			// it will be run via kubernetes cronjob at midnight
```
Comment (Kidswiss) [inline pkg/cmd/spks.go:60]: It uses a ticker further below, so this comment doesn't seem valid anymore.

File: pkg/cmd/spks.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+			sigint := make(chan os.Signal, 1)
+			signal.Notify(sigint, os.Interrupt)
+
+			for loop := true; loop; {
+				select {
+				case <-ticker.C:
+					err = odooClient.SendData(billingRecords)
+					if err != nil {
+						logger.Error(err, "can't export data to Odoo")
+						panic(err)
```
Comment (Kidswiss) [inline pkg/cmd/spks.go:174]: Please don't use panic here and return the error. This way the error is handled by the cli library.

**Review by Kidswiss — CHANGES_REQUESTED** [review]
Just some small things.

File: pkg/cmd/spks.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+				return err
+			}
+
+			odooClient := odoo.NewOdooAPIClient(c.Context, odooURL, odooOauthTokenURL, odooClientId, odooClientSecret, logger)
+			location, err := time.LoadLocation("Europe/Zurich")
+			if err != nil {
+				return fmt.Errorf("load loaction: %w", err)
+			}
+
+			from := time.Now().In(location).UTC()
```
[the remark this answers is not in the document]
Comment (wejdross) [reply pkg/cmd/spks.go]: that's very good catch, I missed it because Odoo simply accepted such payload :D

**Comment by wejdross** [comment]
I added first run of the function before for{} loop to avoid situation in which because of constant restarts we won't send any data

**Comment by Kidswiss** [comment]
> I added first run of the function before for{} loop to avoid situation in which because of constant restarts we won't send any data

Should be fine AFAIK Odoo handles multiple entries for the same timespan gracefully.

**Review by Kidswiss — APPROVED** [review]

File: pkg/cmd/spks.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+		Flags: []cli.Flag{
+			&cli.StringFlag{Name: "odoo-url", Usage: "URL of the Odoo Metered Billing API",
+				EnvVars: []string{"ODOO_URL"}, Destination: &odooURL, Value: "https://preprod.central.vshn.ch/api/v2/product_usage_report_POST"},
+			&cli.StringFlag{Name: "odoo-oauth-token-url", Usage: "Oauth Token URL to authenticate with Odoo metered billing API",
+				EnvVars: []string{"ODOO_OAUTH_TOKEN_URL"}, Destination: &odooOauthTokenURL, Required: true, DefaultText: defaultTextForRequiredFlags},
+			&cli.StringFlag{Name: "odoo-oauth-client-id", Usage: "Client ID of the oauth client to interact with Odoo metered billing API",
+				EnvVars: []string{"ODOO_OAUTH_CLIENT_ID"}, Destination: &odooClientId, Required: true, DefaultText: defaultTextForRequiredFlags},
+			&cli.StringFlag{Name: "odoo-oauth-client-secret", Usage: "Client secret of the oauth client to interact with Odoo metered billing API",
+				EnvVars: []string{"ODOO_OAUTH_CLIENT_SECRET"}, Destination: &odooClientSecret, Required: true, DefaultText: defaultTextForRequiredFlags},
+			&cli.StringFlag{Name: "sales-order", Usage: "Sales order for APPUiO Managed clusters",
```
Comment (tobru) [inline pkg/cmd/spks.go:50]:
```suggestion
			&cli.StringFlag{Name: "sales-order", Usage: "Sales order to report billing data to",
```

This reporting tool isn't about APPUiO Managed

File: pkg/cmd/spks.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+				EnvVars: []string{"SALES_ORDER"}, Destination: &salesOrder, Required: false, DefaultText: defaultTextForOptionalFlags, Value: "S10121"},
+			&cli.StringFlag{Name: "prometheus-url", Usage: "URL of the Prometheus API",
+				EnvVars: []string{"PROMETHEUS_URL"}, Destination: &prometheusURL, Required: true, DefaultText: defaultTextForRequiredFlags, Value: "http://prometheus-monitoring-application.monitoring-application.svc.cluster.local:9090"},
+			&cli.StringFlag{Name: "unit-id", Usage: "Unit ID for the consumed units",
```
Comment (tobru) [inline pkg/cmd/spks.go:54]:
```suggestion
			&cli.StringFlag{Name: "unit-id", Usage: "Metered Billing UoM ID for the consumed units",
```

This is more clear IMHO

File: pkg/cmd/spks.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+			redisPremium, err := QueryPrometheus(ctx, v1api, prometheusQueryArr[3], logger)
+			if err != nil {
+				logger.Error(err, "Error querying Prometheus")
+				return err
+			}
+
+			odooClient := odoo.NewOdooAPIClient(c.Context, odooURL, odooOauthTokenURL, odooClientId, odooClientSecret, logger)
+			location, err := time.LoadLocation("Europe/Zurich")
+			if err != nil {
+				return fmt.Errorf("load loaction: %w", err)
```
Comment (tobru) [inline pkg/cmd/spks.go:100]:
```suggestion
				return fmt.Errorf("load location: %w", err)
```

File: pkg/cmd/spks.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+			from := time.Now().In(location).Add(-(time.Hour * 24)).UTC()
+			to := time.Now().In(location).UTC()
+
+			billingRecords := []odoo.OdooMeteredBillingRecord{
+				{
+					ProductID:            "appcat-spks-mariadb-standard",
+					InstanceID:           "mariadb-standard",
+					ItemDescription:      "appcat-spks-mariadb-standard",
+					ItemGroupDescription: "SPKS",
```
Comment (tobru) [inline pkg/cmd/spks.go:111]: This doesn't make a lot of sense. These two parameters end up in the invoice and should be descriptive to the consumer of the invoice (see https://docs.central.vshn.ch/metered-billing-data-ingestion.html#_json_structure). See example JSON in SPKS-863 for the "ItemDescription". "ItemGroupDescription" should be removed, it's optional anyway.

**Review by tobru — CHANGES_REQUESTED** [review]

**Pull request merged**